python/ray/ml/storage.py

```diff
@@ -0,0 +1,183 @@
+import abc
+import subprocess
+from typing import Dict
+
+from ray import logger
+from ray.util.annotations import DeveloperAPI
+
+S3_PREFIX = "s3://"
+GS_PREFIX = "gs://"
+HDFS_PREFIX = "hdfs://"
+ALLOWED_REMOTE_PREFIXES = (S3_PREFIX, GS_PREFIX, HDFS_PREFIX)
+
+
+class Storage(abc.ABC):
+    """Base class for (external) storage providers.
+
+    Classes inheriting from this provide implementations for
+    methods to upload, download, and delete files and
+    directories to and from external storage.
+    """
+
+    def upload(self, local_source: str, remote_target: str) -> None:
+        """Upload local path to remote target.
+
+        Args:
+            local_source: Path to local source file or directory.
+            remote_target: URI to remote target file or directory.
+        """
+        raise NotImplementedError
+
+    def download(self, remote_source: str, local_target: str) -> None:
+        """Download remote source to local target.
+
+        Args:
+            remote_source: URI to remote source file or directory.
+            local_target: Path to local target file or directory.
+        """
+        raise NotImplementedError
+
+    def delete(self, remote_target: str) -> None:
+        """Delete remote target file or directory..
+
+        Args:
+            remote_target: URI to remote target file or directory.
+        """
+        raise NotImplementedError
+
+
+class S3Storage(Storage):
+    def upload(self, local_source: str, remote_target: str) -> None:
```

Comment:
the sync template has more args like `options` and `exclude_template`. Is this something that we want to support with Storage?

Reply:
I've thought about this, and the problem is that options and excludes are specific to the providers and won't work with others (e.g. HDFS does not support excludes at all). I think for those cases we'll likely drop the exclude_template but extend the S3Storage with tune-specific features. Alternatively, we can support kwargs to enable some of these things.

Reply:
I've added a `**kwargs` parameter to the base API. This will enable us to add arguments in the future.

For Ray Tune, I think we'll have something like

`upload(source, target, sync=True, exclude=["*/checkpoint_*"])` or so.

Different storages can then choose to ignore these parameters (e.g. `sync` likely only applies to S3, and `exclude` only to a select few storages).

With kwargs we're flexible enough to make these changes in the future.